api/envoy/service/rate_limit_quota/v3/rlqs.proto

```diff
@@ -0,0 +1,186 @@
+  // and the server asynchronously responding with the assignments.
+  rpc StreamRateLimitQuotas(stream RateLimitQuotaUsageReports)
+      returns (stream RateLimitQuotaResponse) {
+  }
+}
+
+message RateLimitQuotaUsageReports {
+  // The usage report for a bucket.
+  //
+  // .. note::
+  //   Note that the first report sent for a ``BucketId`` indicates to the RLQS server that
+  //   the RLQS client is subscribing for the future assignments for this ``BucketId``.
+  //
+  // [#next-free-field: 6]
+  message BucketQuotaUsage {
+    // ``BucketId`` for which request quota usage is reported.
+    BucketId bucket_id = 1 [(validate.rules).message = {required: true}];
+
+    // Time elapsed since the last report.
+    google.protobuf.Duration time_elapsed = 2 [(validate.rules).duration = {
+      required: true
+      gt {}
+    }];
+
+    // Requests the data plane has allowed through.
+    uint64 num_requests_allowed = 3;
+
+    // Requests throttled.
+    uint64 num_requests_denied = 4;
+
+    // Requests in the queue waiting for the assignment.
+    uint64 num_requests_queued = 5;
+  }
+
+  // All quota requests must specify the domain. This enables sharing the quota
+  // server between different applications without fear of overlap.
+  // E.g., "envoy".
+  //
+  // Should only be provided in the first report, all subsequent messages on the same
+  // stream are considered to be in the same domain. In case the domain needs to be
+  // changes, close the stream, and reopen a new one with the different domain.
+  string domain = 1 [(validate.rules).string = {min_len: 1}];
+
+  // A list of quota usage reports. The list is processed by the RLQS server in the same order
+  // it's provided by the client.
+  repeated BucketQuotaUsage bucket_quota_usages = 2 [(validate.rules).repeated = {min_items: 1}];
+}
+
+message RateLimitQuotaResponse {
+  message BucketAction {
+    // Quota Assignment for the ``BucketId``. Configures the rate limiting strategy and duration.
+    message QuotaAssignmentAction {
+      // A duration after which the assignment is be considered expired.
+      //
+      // * If unset, the assignment has no expiration date.
+      // * If set to ``0``, the assignment expires immediately, forcing the client into the
+      //   :ref:`"expired assignment"
+      //   <envoy_v3_api_field_extensions.filters.http.rate_limit_quota.v3.RateLimitQuotaBucketSettings.ExpiredAssignmentBehavior.expired_assignment_behavior_timeout>`
+      //   state.
+      google.protobuf.Duration assignment_time_to_live = 2 [(validate.rules).duration = {gte {}}];
+
+      // Configures the local rate limiter for the request matched to the bucket.
+      //
+      // If not set, allow all requests.
+      type.v3.RateLimitStrategy rate_limit_strategy = 3;
+    }
+
+    // Abandon action for the bucket. Indicates that the RLQS server will no longer be
+    // sending updates for this ``BucketId``.
+    message AbandonAction {
+    }
```

Comment:
What should the client do if it gets abandoned? Clarify? I assume go back to expired assignment when the existing assignment expires?

Edit: I see there is detail on this below.

Reply:
Yea, I wasn't where should be the meat of the explanation. @mattklein123  Do you think I should just copy the field description (where all details are) to the message description too? Or maybe move all details to the message, and say "see message description for more details" on the field?

Reply:
I'm fine either way, some cross reference would be good.

Reply:
I revamped this section, see https://github.com/envoyproxy/envoy/pull/19793/commits/699957e2d4a100232c26795468346374acb1c468:

1. Fields `quota_assignment_action` and `abandon_action` now contain minimal information, but they clearly refer to the corresponding messages for the details
2. I moved the existing docs on abandoning the message to `AbandonAction`, and added more details on the procedure of abandoning 
3. I detailed the documentation on applying the assignment in different situations in `QuotaAssignmentAction` messages, and covered some nuances on transitioning between different states.
4. I made other fields in the service and the filter protos refer to the `AbandonAction` and `QuotaAssignmentAction` docs when they talk about abandoning a bucket, or applying the new assignment.

While I was writing it, I realized I initially missed some edge-cases, and missed some implementation details for transitioning between states. So I ended up revamping these sections. There shouldn't be any ground breaking changes, just clarifications on what to do exactly in certain circumstances.